### DISCUSSION TITLE
evolution using slepc for density matrices

### DISCUSSION THREAD

**Discussion by ProThicc** [jcmgray/quimb#207] [General]
Hello,

Is there an expm method implemented for density matrices evolution with slepc backend? I tried passing in a density matrix in the sample mpi code here https://quimb.readthedocs.io/en/latest/distributed%20parallelism%20-%20mpi.html and got this error 

```
[ 6] MFNSolve() at /home/conda/feedstock_root/build_artifacts/slepc_1696510326380/work/src/mfn/interface/mfnsolve.c:90
[ 6] MFNSolve_Private() at /home/conda/feedstock_root/build_artifacts/slepc_1696510326380/work/src/mfn/interface/mfnsolve.c:29
[ 6] Invalid argument
[ 6] Cannot pass a zero b vector to MFNSolve()
```

Here are a few observations while testing it out for n = 5 to figure out the problem
1. 
`gs = qu.up() @ qu.up().H
es = qu.down() @ qu.down().H
psi0 = qu.kron(qu.eye(2),gs,gs,gs,qu.eye(2))/np.sqrt(2**n)`

doesn't show any error

2.
`gs = qu.up() @ qu.up().H
es = qu.down() @ qu.down().H
print(gs)
psi0 = qu.kron(qu.eye(2),gs,es,gs,qu.eye(2))/np.sqrt(2**n)`

shows the error message I posted above.

3.
`gs = qu.up() @ qu.up().H
es = qu.down() @ qu.down().H
print(gs)
psi0 = qu.eye(2**n)

works well

anytime I include "es", this error arises. My guess is when dealing with the parts of psi0 for evolution with MPI, it is encountering a smaller null ket. If anyone has encountered the same issue and were able to bypass it in some way, could you please help me with this?

**Comment by ProThicc** [comment]
I just realised, this is because the method is not defined for density matrices with slepc.

**Reply by jcmgray** [reply]
Hi @ProThicc, yes that is the case. Since the state is generally dense in the density matrix case, there are not iterative methods implemented for it.

**Reply by ProThicc** [reply]
Thank you! Do you know of any other methods implemented for density matrices using mpi?

**Reply by jcmgray** [reply]
You could using something like jump simulation and then the process becomes that of simulating many pure evolutions, but this is not provided directly in `quimb` and its not something i know a lot about!

**Reply by ProThicc** [reply]
That sounds like a good idea! Simulating each quantum Trajectory on each process is viable. Thank you again!
